shelley-ma/formal-spec/value.tex

```diff
@@ -1,77 +1,225 @@
 \section{Coin and Multi-Asset Tokens}
 \label{sec:coin-ma}
 
-This section describes the type $\Value$, the operations required on
-it, and its relation to $\Coin$.
+This ShelleyMA specification allows us to have a single version of a ledger
+spec for both Allegra and Mary eras because we introduce an abstract type,
+$\ValClass$, that allows us to swap out the representation of ledger assets that
+are just Ada (ie. $\Coin$) for a type that represents multiple assets in a
+single term (eg. $\Value$), as well as swap out the type-specific functions
+used in the ledger rules and required for $\ValClass$ instantiation (see Figure \ref{fig:valclass}).
+We do so by instantiating $\ValClass$ to either one of the types. Note that
+it can, in theory, be instantiated to any other type on which we can meaningfully
+define the required functions (addition, size, etc.).
+
+\subsection{$\ValClass$}
+
+$\ValClass$ is intuitively used as a placeholder for any type that can be used
+to represent (inside a transaction or in the ledger UTxO) quantities of a single asset or several different assets.
+Figure \ref{fig:valclass} gives names and types of functions in which $\ValClass$ is either
+an argument or the output type, and so they cannot be defined without assigning
+the abstract $\ValClass$ a specific type. Because these functions are used in the
+Allegra and Mary transition rules, to define the two concrete ledger models,
+we must not only pick a type with which to instantiate the $\ValClass$ abstraction,
+but also give concrete definitions to the associated functions.
+
+
+%%
+%% Figure ValClass and its Functions
+%%
+\begin{figure}[htb]
+  \emph{Helper function}
+  %
+  \begin{align*}
+    \mathsf{ValType} &\in \ValClass  \\
```

Comment:
Why are there two definitions of `ValType`? Also, `ValType \in \ValClass` isn't really correct

Reply:
This also has the wrong alignment in the PDF output

Reply:
do you have any idea how to better specify the relation between `ValType` and `ValClass`?

Reply:
See the comment above about the more traditional way of defining it. For a type class, it could either be a parameter of the class or a member of it. In the former case it's probably best to mention in in the prose and call it an `Abstract type`, and in the latter we could just say `ValType \in \Type` in this figure.